File: QytetetJava/src/modeloqytetet/TituloPropiedad.java
////////////////////////////////////////////////////////////////////////////////
// Autor: Alejandro Manzanares Lemus
// 2º GII A1
////////////////////////////////////////////////////////////////////////////////
package modeloqytetet;
import java.util.logging.Logger;
    
public class TituloPropiedad {
    //Atributos
    private int alquilerBase;
    private float factorRevalorizacion;
    private int hipotecaBase;
    private boolean hipotecada;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private int precioCompra;
    private int precioEdificar;
    private Jugador propietario;
    
    //Constructor
    public TituloPropiedad(String nom, int precioC, int alquiler, 
            float factor, int precioE, int hipoteca){
        hipotecada = false;
        numHoteles = 0;
        numCasas = 0;
        nombre = nom;
        precioCompra = precioC;
        alquilerBase = alquiler;
        factorRevalorizacion = factor;
        hipotecaBase = hipoteca;
        precioEdificar = precioE;
        propietario = null;
    }
    
    //Metodos
    int calcularCosteCancelar(){
        return (int) (calcularCosteHipotecar() + calcularCosteHipotecar() * 0.1);
    }
    int calcularCosteHipotecar(){
        int costeHipoteca = (int) (hipotecaBase + (numCasas*0.5*hipotecaBase+numHoteles*hipotecaBase));
        return costeHipoteca;
    }
    int calcularImporteAlquiler(){
        int costeAlquiler = (int)(alquilerBase + (numCasas*0.5+numHoteles*2));
        return costeAlquiler;
    }
    int calcularPrecioVenta(){
        int precioVenta = (int) (precioCompra + (numCasas+numHoteles) *precioEdificar *factorRevalorizacion);
        return precioVenta;
    }
    void cancelarHipoteca(){
        hipotecada = false;
    }
    void edificarCasa(){
        numCasas += 1;
    }
    void edificarHotel(){
        numHoteles += 1;
    }
    String getNombre() {
        return nombre;
    }
    boolean isHipotecada() {
        return hipotecada;
    }
    int getPrecioCompra() {
        return precioCompra;
    }
    int getAlquilerBase() {
        return alquilerBase;
    }
    float getFactorRevalorizacion() {
        return factorRevalorizacion;
    }
    int getHipotecaBase() {
        return hipotecaBase;
    }
    int getPrecioEdificar() {
        return precioEdificar;
    }
    int getNumCasas() {
        return numCasas;
    }
    int getNumHoteles() {
        return numHoteles;
    }
    Jugador getPropietario(){
        return propietario;
    }    
    int hipotecar(){
        hipotecada = true;
        int costeHipoteca = calcularCosteHipotecar();
        return costeHipoteca;
    }
    int pagarAlquiler(){
       int costeAlquiler = calcularImporteAlquiler();
       propietario.modificarSaldo(costeAlquiler);
       return costeAlquiler;
    }
    boolean propietarioEncarcelado(){
        return propietario.getEncarcelado();
    }
    boolean tengoPropietario(){
        return (propietario != null);
    }
    void setHipotecada(boolean hipoteca) {
        hipotecada = hipoteca;
    }
    void setPropietario(Jugador propietario){
        this.propietario = propietario;
    }
    @Override
    public String toString( ){
        String nombreprop;
        if(propietario != null){
            nombreprop = propietario.getNombre();
        }
        else{
            nombreprop = "No hay propietario";
        }
        return "TituloPropiedad{" + "\n"
            + " nombre=" + nombre + "\n"
            + " nombre propietario=" + nombreprop + "\n"
            + " hipotecado=" + Boolean.toString(hipotecada) + "\n"
            + " precioCompra=" + Integer.toString(precioCompra) + "\n"
            + " alquilerBase=" + Integer.toString(alquilerBase) + "\n"
            + " factorRevalorizacion=" + Float.toString(factorRevalorizacion)
                + "\n"
            + " hipotecaBase=" + Integer.toString(hipotecaBase) + "\n"
            + " precioEdificar=" + Integer.toString(precioEdificar) + "\n"
            + " numCasas=" + Integer.toString(numCasas) + "\n"
            + " numHoteles=" + Integer.toString(numHoteles)
            + "}";
    }

    //No salen en el diagrama
    void cobrarAlquiler(int coste){
        throw new UnsupportedOperationException("Sin implementar");
    }
}   

File: QytetetRuby/lib/titulo_propiedad.rb
#///////////////////////////////////////////////////////////////////////////////
#// Autor: Alejandro Manzanares Lemus
#// 2º GII A1
#///////////////////////////////////////////////////////////////////////////////
module ModeloQytetet
  class TituloPropiedad
    def initialize(nom, precioC, alquiler, factor, precioE, hipoteca)
        @nombre = nom
        @hipotecada = false
        @precio_compra = precioC
        @alquiler_base = alquiler
        @factor_revalorizacion = factor
        @hipoteca_base = hipoteca
        @precio_edificar = precioE
        @num_casas = 0
        @num_hoteles = 0
        @propietario = nil
    end

    attr_reader :nombre, :precio_compra, :alquiler_base, :factor_revalorizacion,
      :hipoteca_base, :precio_edificar, :num_casas, :num_hoteles
    attr_accessor :hipotecada, :propietario
    
    def calcular_coste_cancelar()
      coste_cancelar = (calcular_coste_hipotecar() + calcular_coste_hipotecar() *0.1).to_i
    end
    def calcular_coste_hipotecar()
      coste_hipoteca = @hipoteca_base + (@num_casas * 0.5 * @hipoteca_base + @num_hoteles * @hipoteca_base)#.to_i
    end
    def calcular_importe_alquiler()
      coste_alquiler = @alquiler_base + (@num_casas*0.5 + @num_hoteles*2).to_i
    end
    def calcular_precio_venta()
      precio_venta = @precio_compra + (@num_casas + @num_hoteles) * @precio_edificar * @factor_revalorizacion
    end
    def cancelar_hipoteca()
      @hipotecada = false
    end
    def cobrar_alquiler(coste)
      raise NotImplementedError
    end
    def edificar_casa()
      @num_casas +=1
    end
    def edificar_hotel()
      @num_hoteles +=1
    end
    def hipotecar()
      @hipotecada = true
      coste_hipoteca = calcular_coste_hipotecar
    end
    def pagar_alquiler()
      coste_alquiler = calcular_importe_alquiler()
      @propietario.modificar_saldo(coste_alquiler)
    end
    def propietario_encarcelado()
      return @encarcelado
    end
    def tengo_propietario()
      return !@propietario.nil?
    end
    def to_s()
      if (@propietario.nil?)
        nombreprop = "Nadie"
      else 
        nombreprop = @propietario.nombre
      end
      "TITULO_PROPIEDAD:
Nombre:#{@nombre}
Propietario:#{nombreprop}
Hipotecada:#{@hipotecada}
Precio_compra:#{@precio_compra}
Alquiler_base:#{@alquiler_base}
Factor_revalorizacion:#{@factor_revalorizacion}
Hipoteca_base:#{@hipoteca_base}
Precio_edificar:#{@precio_edificar}
Num_casas:#{@num_casas}
Num_hoteles:#{@num_hoteles}"
    end
  end
end